### PR TITLE
Interface Sparse BLAS routines

### PR DESCRIPTION
**Pull request by amontoison**
@AntoninKns

### CODE REVIEW

**Comment by geoffroyleconte** [comment]
Hi, thanks for the PR. Could you indicate in the documentation how to use these routines, and add tests if possible?

**Comment by amontoison** [comment]
> Hi, thanks for the PR. Could you indicate in the documentation how to use these routines, and add tests if possible? I think that the old `mul!` implementation is still tested.

Hi @geoffroyleconte, I didn't finished the PR but you don't to update something to use the new routine.
The `mul!` will dispatch to multithreaded SparseBLAS MKL routines when it can.
I will add new tests but the current tests should use the new routines.

File: src/SparseMatricesCOO.jl [hunk 1]
```diff
@@ -11,8 +11,19 @@ using LinearAlgebra, SparseArrays
 # this should be revisited after https://github.com/JuliaLang/julia/pull/33821
 using UnicodePlots
 
+using MKL_jll
+
+function __init__()
+  ccall((:MKL_Set_Interface_Layer, libmkl_rt), Cint, (Cint,), Base.USE_BLAS64 ? 1 : 0)
```
Comment (amontoison) [inline src/SparseMatricesCOO.jl:17]:
```suggestion
  if Sys.islinux() || Sys.isapple() || Sys.iswindows()
    ccall((:MKL_Set_Interface_Layer, libmkl_rt), Cint, (Cint,), Base.USE_BLAS64 ? 1 : 0)
  end
```

**Comment by geoffroyleconte** [comment]
> > Hi, thanks for the PR. Could you indicate in the documentation how to use these routines, and add tests if possible? I think that the old `mul!` implementation is still tested.
> 
> Hi @geoffroyleconte, I didn't finished the PR but you don't need to update something to use the new routines. The `mul!` will dispatch to multithreaded SparseBLAS MKL routines when it can. I will add new tests and benchmarks but the current tests should use the new routines.
> 
> It will be more efficient on Intel CPUs but we could still have a significant speed-up on AMD / M1 processors because the routines are multithreaded.

Perfect thanks! Yes I think most of the tests use `mul!` instead of `*` so that might be why there are some coverage issues

**Comment by dpo** [comment]
> I will add new tests and benchmarks but the current tests should use the new routines.

They should or they do?

**Comment by amontoison** [comment]
> > I will add new tests and benchmarks but the current tests should use the new routines.
> 
> They should or they do?

They do, I checked with `@code_warntype`.

File: src/coo_mkl_wrapper.jl [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+function check_uplo(uplo::Char)
+  uplo == 'L' ||  uplo == 'U' || error("uplo is '$uplo', must be L' or 'U'")
+end
+
+function check_diag(diag::Char)
+  diag == 'U' ||  diag == 'N' || error("diag is '$diag', must be U' or 'N'")
+end
+
+for (mv, sv, symv, trsv, mm, sm, T) in (("mkl_scoomv", "mkl_scoosv", "mkl_scoosymv", "mkl_scootrmv", "mkl_scoomm", "mkl_scoosm", :Float32),
```
Comment (dpo) [inline src/coo_mkl_wrapper.jl:24]: This is likely to get all messed up by the formatter. Better use noformat blocks.

**Comment by amontoison** [comment]
Time in seconds to perform 1000 matrix-vector products with matrices from the SuiteSparseMatrixCollection.
It seems that the MKL routine for COO matrices is not multithreaded :(
Capture d’écran du 2022-10-25 02-06-27

**Comment by geoffroyleconte** [comment]
It's not much faster. Maybe SparseCOO matrix-vector products cannot be multi-threaded? Even in Julia I do not know how we could use the `@threads` macro because we do not know the number of nonzeros in each column.
Are the benchmark performed on matrices with the `Symmetric` wrapper? I'm not sure that this function is the fastest possible with the `t` call : https://github.com/JuliaSmoothOptimizers/SparseMatricesCOO.jl/blob/c81dc762eb39f3a60b58163edd72eafc20558aaf/src/coo_linalg.jl#L66-L75

**Comment by amontoison** [comment]
We have the same performances for Symmetric matrices.
benchmarks_sym_coo

**Comment by dpo** [comment]
Not exactly in favor of MKL. What's going on here?

**Comment by dpo** [comment]
> I'm not sure that this function is the fastest possible with the t call :

A little meta-programming might settle that.

**Comment by amontoison** [comment]
> Not exactly in favor of MKL. What's going on here?

I interfaced deprecated routines ([coo_mv](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/blas-and-sparse-blas-routines/sparse-blas-level-2-and-level-3-routines/sparse-blas-level-2-and-level-3-routines-1/mkl-coomv.html)).
I should try the new and generic `mkl_sparse_mv`.
I will need Clang.jl to generate the interface.

**Comment by amontoison** [comment]
I interfaced the new routines, the MKL version is slower than our Julia version. 
benchmarks_coo_png
benchmarks_sparse_matrices_png

**Comment by dpo** [comment]
who needs the MKL? 😃

**Comment by amontoison** [comment]
I close the PR. MKL is not relevant for COO.
I just added a dummy copy of our COO sparse matrix in `MKLSparse.jl` to easily test if the performance are better in the future:
https://github.com/JuliaSparse/MKLSparse.jl/pull/33/files#diff-6241561dad231b63a9c2e46e7ff91ecfb885c834a98b621364ab11d323086af7R1-R19

**Pull request closed without merging**